### PR TITLE
bpo-41796: Make _ast module state per interpreter

### PR DESCRIPTION
**Pull request by vstinner**
The ast module internal state is now per interpreter.

* Rename "astmodulestate" to "struct ast_state"
* Add pycore_ast.h internal header: the ast_state structure is now
  declared in pycore_ast.h.
* Add PyInterpreterState.ast (struct ast_state)
* Remove get_ast_state()
* Rename get_global_ast_state() to get_ast_state()
* PyAST_obj2mod() now handles get_ast_state() failures

https://bugs.python.org/issue41796

### CODE REVIEW

**Comment by vstinner** [comment]
In short, this change replaces `global_ast_state` of Python-ast.c with PyInterprreterState.ast. It removes:

```
// bpo-41194, bpo-41261, bpo-41631: The _ast module uses a global state.
static astmodulestate global_ast_state = {0};
```

cc @encukou @pablogsal @lysnikolaou

**Comment by vstinner** [comment]
Oh, I missed test_peg_generator. I should now be fixed with my second commit.

**Review by lysnikolaou — COMMENTED** [review]
Should the Windows regen file (`regen.vcxproj`) be updated as well?

File: PCbuild/regen.vcxproj [hunk 1]
```diff
@@ -161,15 +163,18 @@
     <Warning Text="Pegen updated. You will need to rebuild pythoncore to see the changes." Condition="'@(_UpdatedParse)' != ''" />
   </Target>
   <Target Name="_RegenAST_H" AfterTargets="_RegenGrammar">
-    <!-- Regenerate Include/Python-ast.h and Python/Python-ast.c using Parser/asdl_c.py -h -->
-    <Exec Command="&quot;$(PythonExe)&quot; &quot;$(PySourcePath)Parser\asdl_c.py&quot; &quot;$(PySourcePath)Parser\Python.asdl&quot; -H &quot;$(IntDir)Python-ast.h&quot; -C &quot;$(IntDir)Python-ast.c&quot;" />
+    <!-- Regenerate Include/Python-ast.h, Python/Python-ast.c annd Include/internal/pycore_ast.h using Parser/asdl_c.py -h -->
```
Comment (isidentical) [inline PCbuild/regen.vcxproj]:
```suggestion
    <!-- Regenerate Include/Python-ast.h, Python/Python-ast.c and Include/internal/pycore_ast.h using Parser/asdl_c.py -h -->
```

File: Parser/asdl_c.py [hunk 2]
```diff
@@ -1341,7 +1341,35 @@ def visit(self, object):
...
+        f.write('    PyObject *' + s + ';\n')
+    f.write('};')
+
+
+def generate_ast_fini(module_state, f):
+    f.write("""
+void _PyAST_Fini(PyThreadState *tstate)
+{
+#ifdef Py_BUILD_CORE
+    struct ast_state* state = &tstate->interp->ast;
```
Comment (isidentical) [inline Parser/asdl_c.py]:
```suggestion
    struct ast_state *state = &tstate->interp->ast;
```

File: Parser/asdl_c.py [hunk 3]
```diff
@@ -1093,12 +1094,7 @@ def visitModule(self, mod):
         self.emit("static int", 0)
         self.emit("astmodule_exec(PyObject *m)", 0)
         self.emit("{", 0)
-        self.emit('astmodulestate *state = get_ast_state(m);', 1)
-        self.emit("", 0)
-
-        self.emit("if (!init_types(state)) {", 1)
-        self.emit("return -1;", 2)
-        self.emit("}", 1)
+        self.emit('struct ast_state *state = get_ast_state();', 1)
```
Comment (isidentical) [inline Parser/asdl_c.py:1097]: I guess we need to check for `state == NULL` here (in case of init_types failed)

Comment (vstinner) [reply Parser/asdl_c.py:1097]: Oh right, fixed, thanks.

**Comment by vstinner** [comment]
@isidentical: Thanks for the review. I addressed your review. Please review the updated PR.

**Comment by vstinner** [comment]
@lysnikolaou: "Should the Windows regen file (regen.vcxproj) be updated as well?"

Oh, I didn't know this project. Thanks, I updated it.

**Comment by vstinner** [comment]
I merged my PR. Thanks for the useful review @lysnikolaou and @isidentical!

I dislike having to declare the struct ast_state once in pycore_ast.h and once in Python-ast.c, but Python-ast.c is a strange beast which is used outside Python without Py_BUILD_CORE. So declaring the structure twice is a practical solution for that.

@encukou used a different solution: use a pointer to an opaque structure in the PyInterpreterState, and allocate the structure on the heap memory when it's initialized. That would avoid to declare the structure twice. But I chose to follow the current trend of declaring all structures in pycore_interp.h and have a large unique and **flat** PyInterpreter structure (no pointer indirection).

Maybe tomorrow we should revisit this design to have something more flexible, to not have to declare everything in pycore_interp.h.

The best would be to be able to have a real module state, but I was biten multiple times by the C API which requires the AST module state, whereas the C API isn't directly related to a module. I tried a solution using an import and we got a crash with Mercurial lazy import, etc.

**Comment by pablogsal** [comment]
I am relatively sure that this commit introduced reference leaks in the test suite:

https://buildbot.python.org/all/#/builders/205/builds/83

OK
......
test_ast leaked [23640, 23636, 23640] references, sum=70916
test_ast leaked [7932, 7930, 7932] memory blocks, sum=23794
1 test failed again:
    test_ast

See also https://bugs.python.org/issue42250

**Pull request merged**